// Copyright 2024 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

syntax = "proto3";

package flwr.proto;

import "flwr/proto/heartbeat.proto";
import "flwr/proto/log.proto";
import "flwr/proto/message.proto";
import "flwr/proto/run.proto";
import "flwr/proto/fab.proto";
import "flwr/proto/appio.proto";

service SimulationIo {

  // List runs to launch
  rpc ListAppsToLaunch(ListAppsToLaunchRequest)
      returns (ListAppsToLaunchResponse) {}

  // Request token for a run
  rpc RequestToken(RequestTokenRequest) returns (RequestTokenResponse) {}

  // Pull Simulation inputs
  rpc PullAppInputs(PullAppInputsRequest) returns (PullAppInputsResponse) {}

  // Push Simulation outputs
  rpc PushAppOutputs(PushAppOutputsRequest) returns (PushAppOutputsResponse) {}

  // Update the status of a given run
  rpc UpdateRunStatus(UpdateRunStatusRequest)
      returns (UpdateRunStatusResponse) {}

  // Push ServerApp logs
  rpc PushLogs(PushLogsRequest) returns (PushLogsResponse) {}

  // Get Federation Options
  rpc GetFederationOptions(GetFederationOptionsRequest)
      returns (GetFederationOptionsResponse) {}

  // Get Run Status
  rpc GetRunStatus(GetRunStatusRequest) returns (GetRunStatusResponse) {}

  // Heartbeat
  rpc SendAppHeartbeat(SendAppHeartbeatRequest)
      returns (SendAppHeartbeatResponse) {}
}
